### PR TITLE
fix: delete message action logic

### LINKED ISSUE #14488: Edge case when delete pinned messages in community
This is a follow-up issue for https://github.com/status-im/status-mobile/pull/14232

Need input from designer. Q: what to do when 
1. user A sent a message
2. admin pinned the message
3. user A try to delete the message while don't have the permission to unpin the message

Potential solutions
1. user A can delete the message, the pinned message turns to "message deleted by user xxx" ✅
2. ~user A can't delete the message, hide the delete message for everyone option in the context menu for such kind of messages~

### LINKED ISSUE #14510: The message is not shown as deleted in the pinned section if 'delete for everyone' option is used
**Steps to reproduce:**
1. Go to the 1-1 chat
2. Send the message: "Test_A"
3. Leave a reply: "Test_B"
4. Pin the message
5. Remove replied message "Test_A"
6. Open the list of pinned messages
7. Remove the pinned message using 'delete for everyone' option
8. Check the replied message

**Actual result:**
The replied message "Test_A" is shown after removing

https://user-images.githubusercontent.com/52490791/214904661-2b3cea83-a4c0-404d-a67d-8acd01d516b9.mp4

image

**Expected result:**
The pinned message is shown as removed in the chat
image

### LINKED ISSUE #14729: It is possible to pin already pinned message
**Steps:**

1. Open any chat
2. Send a message
3. Pin that message
4. Close chat
5. Re enter the chat
6. Open list of pinned messages by tapping on pinned banner
7. Longtap on pinned message
8. See if unpin option is displayed for that message

**Actual result:** pin instead of unpin option is shown. It is possible perform pin action for already pinned message

**Expected result:** unpin option is displayed

https://user-images.githubusercontent.com/97245802/211273075-cb94c1d8-7d3c-47b0-8506-1e0bc64d00ca.mp4

### Additional Information

- Status version: nightly 
- Operating System: Android, iOS

[comment]: # (Please, add logs/notes if necessary)

### LINKED ISSUE #14938: No delete for everyone option in group chat for group admin
# Bug Report
in group chat, there's no delete for everyone option for group admin when long press on others message

### LINKED ISSUE #15034: Some components disappear from the list of pinned messages after chat reopening
**Steps to reproduce:**
1. Go to the 1-1 chat
2. Send a message
3. Pin the message
4. Reopen the chat

**Actual result:**
The following info is dissapaered:
- Owner and his chatkey
- Timemark
- Pin
- Style for pinned messages

https://user-images.githubusercontent.com/52490791/217331367-b65df057-1dd0-43ce-baeb-5cdddedd8dd8.mp4

**Expected result:**
image

**ENV:**
- Nightly 7 Feb 2023 [2619]

### PR DESCRIPTION
**Pull request by yqrashawn**
fix https://github.com/status-im/status-mobile/issues/14938
fix https://github.com/status-im/status-mobile/issues/14510 
fix https://github.com/status-im/status-mobile/issues/14488
fix https://github.com/status-im/status-mobile/issues/14729
fix https://github.com/status-im/status-mobile/issues/15034

1. ~rename `:can-delete-message-for-everyone` to `:can-delete-message-for-everyone-in-community?`~
1. use same logic for actions in message list and actions in pinned message list
1. do not filter out deleted messages for pinned message list
1. add deleted message UI for pinned message list
1. add drawer for users who have the permission to unpin message for deleted pinned message
1. re pin message when undo pinned message deletes
1. add `:pinned` for messages in pinned message list, this resolves several issues where there are differences between message ui/logic in chat and pinned message list (#14729 #15034)

Note: currently delete message action will always try to unpin the message, relies on status-go to check if user has the permission to unpin message
image

status: ready

### CODE REVIEW

File: src/status_im2/contexts/chat/messages/pin/list/view.cljs [hunk 1]
```diff
@@ -1,37 +1,56 @@
...
+     :community?                                    (not (nil? community-id))
+     :group-admin?                                  group-admin?
+     :community-admin?                              community-admin?
+     :current-public-key                            current-public-key
+     :space-keeper                                  space-keeper
+     :chat-id                                       chat-id
+     :message-pin-enabled                           message-pin-enabled
+     :edit-enabled                                  edit-enabled
+     :in-pinned-view?                               in-pinned-view?
+     :can-delete-message-for-everyone-in-community? can-delete-message-for-everyone-in-community?}))
```
Comment (yqrashawn) [inline src/status_im2/contexts/chat/messages/pin/list/view.cljs]: Copied from `status-im2.contexts.chat.messages.list.view`
I am uncertain if it better to copy or reuse the function
https://github.com/status-im/status-mobile/blob/def02d48dabc7996b9a4d7ffd1f28fbf8ff46838/src/status_im2/contexts/chat/messages/list/view.cljs#L71

**Comment by yqrashawn** [comment]
back to draft, wrong wrong UI included

**Comment by yqrashawn** [comment]
ready for review

File: src/status_im/communities/core.cljs [hunk 2]
```diff
@@ -68,7 +68,7 @@
   (-> c
       (set/rename-keys {:canRequestAccess            :can-request-access?
                         :canManageUsers              :can-manage-users?
-                        :canDeleteMessageForEveryone :can-delete-message-for-everyone?
+                        :canDeleteMessageForEveryone :can-delete-message-for-everyone-in-community?
```
Comment (flexsurfer) [inline src/status_im/communities/core.cljs]: could you elaborate? this map is for the community, why do we need to specify `community` for the specific key ?

File: src/status_im2/contexts/chat/messages/drawers/view.cljs [hunk 3]
```diff
@@ -60,7 +62,13 @@
        :label    (i18n/label :t/delete-for-me)
        :icon     :i/delete
        :id       :delete-for-me}])
-   (when (and (or outgoing can-delete-message-for-everyone?) config/delete-message-enabled?)
+   (when (and config/delete-message-enabled?
```
Comment (flexsurfer) [inline src/status_im2/contexts/chat/messages/drawers/view.cljs]: why do we need config condition?

File: src/status_im2/contexts/chat/messages/pin/list/view.cljs (hunk 1 above)
Comment (flexsurfer) [reply src/status_im2/contexts/chat/messages/pin/list/view.cljs]: better to move it and reuse

File: src/status_im2/contexts/chat/messages/drawers/view.cljs [hunk 4]
```diff
@@ -60,7 +62,13 @@
...
+              (cond
+                outgoing   true
+                community? (or can-delete-message-for-everyone-in-community?
+                               community-admin?)
```
Comment (qfrank) [inline src/status_im2/contexts/chat/messages/drawers/view.cljs]: I believe can-delete-message-for-everyone-in-community? already contains logical that judge if user is admin in status-go side?

File: src/status_im/communities/core.cljs (hunk 2 above)
Comment (yqrashawn) [reply src/status_im/communities/core.cljs]: Fixed, I thought it's under message

File: src/status_im2/contexts/chat/messages/drawers/view.cljs (hunk 4 above)
Comment (yqrashawn) [reply src/status_im2/contexts/chat/messages/drawers/view.cljs]: Fixed

File: src/status_im2/contexts/chat/messages/drawers/view.cljs (hunk 3 above)
Comment (yqrashawn) [reply src/status_im2/contexts/chat/messages/drawers/view.cljs]: Seems this is the only place that uses this config. 
I'm not sure what this config does.
It's introduced at https://github.com/status-im/status-mobile/pull/12491
Seems we can safely delete this.

Comment (yqrashawn) [reply src/status_im2/contexts/chat/messages/drawers/view.cljs]: removed

**Review by flexsurfer — APPROVED** [review]

File: src/status_im2/contexts/chat/messages/pin/list/view.cljs (hunk 1 above)
Comment (yqrashawn) [reply src/status_im2/contexts/chat/messages/pin/list/view.cljs]: fixed by reuse the one in `status-im2.contexts.chat.messages.list.view`
added doc string `"compute data used to render message list, including pinned message list and message list in chats"` for `get-render-data`

**Comment by yqrashawn** [comment]
This PR makes it possible to leave deleted messages in pinned message list. There's no way to delete them since there's no drawer for deleted message. 
Created a follow-up issue for this https://github.com/status-im/status-mobile/issues/14998

**Review by qfrank — APPROVED** [review]

**Comment by pavloburykh** [comment]
### QA NOTE: 
Check fixing of this crash https://github.com/status-im/status-mobile/pull/14965#issuecomment-1416026582

**Comment by VolodLytvynenko** [comment]
Hi @yqrashawn . could you rebase this PR, please?

**Comment by yqrashawn** [comment]
> Hi @yqrashawn . could you rebase this PR, please?

rebased

**Comment by VolodLytvynenko** [comment]
Hi @yqrashawn thank you for PR. Here is the list of found issues:

**ISSUE 1: The pinned message text is not shown in the 'pin banner'**
**Steps to reproduce:**
1. Go to 1-1 chat 
2. Send the message
3. Pin the message
4. Check 'pin panner'

**Actual result:**
An empty pin banner is shown

https://user-images.githubusercontent.com/52490791/217290291-be28bc68-f6db-4266-9e65-d4c1ffe1d04c.mp4


**Expected result:**
The message content is shown in the 'pin banner'
image

**Additional info:**
This issue is related to group chat and community as well

**ISSUE 2 The removed message is not shown as removed in the list of pinned messages**
**Steps to reproduce:**
1. Go to any chat
2. Send the message
3. Pin the message
4. Remove the message with the 'remove for everyone' option
5. Wait 3-4 seconds
6. Check the list of pinned messages 

**Actual result:**
The message is not shown within the list of pinned messages as removed

https://user-images.githubusercontent.com/52490791/217298190-96eb9e80-9b51-4391-a879-5dcb597be4d3.mp4

**Expected result:**
The removed message is shown as removed within the list of pinned messages
image

**ISSUE 3: The owner who removed the message is not shown after the non-admin removes the pinned message and reopens the chat**
**Preconditions:**
- User_A is admin of the community
- User_A and User_b are members of the community

**Steps to reproduce:**
1. Go to the community (USER_B)
2. Send the message  (USER_B)
3. Pin the message (USER_A)
4. Remove the pinned  message (USER_B)
5. Reopen the chat
6. Open the list of pinned messages

**Actual result:**
Owner of the removed message is not shown
image


**Expected result:**
Owner of the removed message is shown within the list of pinned messages

**Comment by VolodLytvynenko** [comment]
> 1. rename `:can-delete-message-for-everyone` to `:can-delete-message-for-everyone-in-community?`

Hi @yqrashawn I didn't understand this row. Is it testable :) or this is some changes related to the code?

**Comment by VolodLytvynenko** [comment]
I suppose ISSUE 3 will be fixed here https://github.com/status-im/status-mobile/issues/15034

**Comment by yqrashawn** [comment]
> > 1. rename `:can-delete-message-for-everyone` to `:can-delete-message-for-everyone-in-community?`
> 
> Hi @yqrashawn I didn't understand this row. Is it testable :) or this is some changes related to the code?

@VladimrLitvinenko 

This is just renaming variables, and it's reverted.

ISSUE 1 is fixed

ISSUE 2 is the desired behavior. The logic behind this is when the user deletes a message for everyone, the user will also unpin the message if it's pinned and if the user has the permission to unpin the message. 
Pinned deleted message only exists when a user deletes a pinned message without the unpin permission.

I'll create a separate PR for ISSUE3

**Comment by yqrashawn** [comment]
Found another issue, working on it

when user without unpin permission delete the message, the message will disappear in pinned message list (and pinned message banner) and will reappear when reenter the chat view

**Comment by VolodLytvynenko** [comment]
> Found another issue, working on it
> 
> when user without unpin permission delete the message, the message will disappear in pinned message list (and pinned message banner) and will reappear when reenter the chat view

@yqrashawn Thanks! Let me know when PR will be ready for tests

**Comment by yqrashawn** [comment]
Updates:
1. add drawer for users who have the permission to unpin message for deleted pinned message
1. re pin message when undo pinned message deletes
1. add `:pinned` for messages in pinned message list, this resolves several issues where there are differences between message ui/logic in chat and pinned message list (#14729 #15034)

Working on:
sync client-side state (state before reenter chat view) of messages in pinned message list and chat view, eg. state of message in pinned message list when delete pinned message and undo immediately

@status-im/mobile-qa

**Comment by yqrashawn** [comment]
> state of message in pinned message list when delete pinned message and undo immediately

this one is fixed

re request code review since lots has changed @flexsurfer @qfrank

**Comment by qoqobolo** [comment]
@yqrashawn thanks for the updates!
Kindly ping us again once the PR gets approval from the reviewers 🙏

Also, please note that the whole QA team is now busy testing the release, so PRs not included in the rc1 scope will be tested after the release testing is completed.
Thank you for your understanding and waiting!

File: src/status_im2/contexts/chat/messages/pin/list/view.cljs [hunk 5]
```diff
@@ -1,37 +1,34 @@
...
+    [content.deleted/deleted-message message context]
+    [message/message-with-reactions message context]))
 
 (defn pinned-messages-list
   [chat-id]
-  (let [pinned-messages (vec (vals (rf/sub [:chats/pinned chat-id])))
-        current-chat    (rf/sub [:chat-by-id chat-id])
-        community       (rf/sub [:communities/community (:community-id current-chat)])]
+  (let [pinned-messages                                          (->> (rf/sub [:chats/pinned chat-id])
+                                                                      vals
```
Comment (flexsurfer) [inline src/status_im2/contexts/chat/messages/pin/list/view.cljs]: can this be done in subscription?

**Review by flexsurfer — APPROVED** [review]

File: src/status_im2/contexts/chat/messages/pin/list/view.cljs (hunk 5 above)
Comment (yqrashawn) [reply src/status_im2/contexts/chat/messages/pin/list/view.cljs]: Fixed using the already defined `:chats/pinned-sorted-list`

**Review by qfrank — APPROVED** [review]

**Comment by VolodLytvynenko** [comment]
Hi @yqrashawn Thanx for PR. The 1-2 issues are fixed. Here is the list of found issues:
## ISSUE 4: “Null is not an object”  error is shown when a message is removed by deleted for everyone option

**Steps to reproduce:**
1. User_A and User_B are mutual contacts
2. Go to 1-1 chat (1-1 chat is just an example. it is reproducible for all chats)
3. Send a message
4. Remove the message by 'delete for everyone' option
5. Wait at least 3-5 seconds

**Actual result:**
"Null is not an object" error  is shown after the message is removed

https://user-images.githubusercontent.com/52490791/219353939-97e0dbaf-0ffd-4af2-9eef-607d6e285252.mp4


**Expected result:**
The error is not shown after the message is removed

**Env:** IOS and Androids

**Logs:** 
[geth.log](https://github.com/status-im/status-mobile/files/10755461/geth.log)
[Status.log](https://github.com/status-im/status-mobile/files/10755466/Status.log)

**Comment by pavloburykh** [comment]
Hi @yqrashawn !
I see that you mentioned this one https://github.com/status-im/status-mobile/issues/14729 among fixes of this PR. 
But this issue has already been fixed by https://github.com/status-im/status-mobile/pull/14787 and merged into develop.

**Comment by yqrashawn** [comment]
> Hi @yqrashawn ! I see that you mentioned this one #14729 among fixes of this PR. But this issue has already been fixed by #14787 and merged into develop.

Yes, the fix was added before #14787 merged. The approach in this PR managed to fix both #14729 and #15034, they actually caused by the same issue which is `:pinned true` exists in chat list msg but missing in pinned message list msg. So have to fix them both.

**Comment by VolodLytvynenko** [comment]
## ISSUE 5: The message is not shown as removed for a non-admin user in the list of pinned messages unless the chat is reopened

**Steps to reproduce:**
1. Open community chat as User_B (non-admin) -> send the message
2. Pin the message using User_B (admin)
3. Remove the message as User_B
4. Check pin banner User_B

**Actual result:**
The pin banner is disappeared (will appear again only if the chat is reopened)

https://user-images.githubusercontent.com/52490791/219393439-28dcb47e-7c87-432f-9644-b425006e1b13.mp4

**Expected result:**
The banner is shown if the message is removed and the current message is shown there as removed


## ISSUE 6: The content of a message is shown in the pin banner for an admin user after a current message is removed by a non-admin user

**Steps to reproduce:**
1. Open community chat as User_B (non-admin) -> send the message
2. Pin the message using User_B (admin)
3. Remove the message as User_B
4. Check pin banner as User_A

**Actual result:**
The message is shown for admin user with text content in the list of pinned
https://user-images.githubusercontent.com/52490791/219393360-550e88d2-79e5-46fd-892c-14d2eaad8e0c.mp4

**Expected result:**
The message is shown for the admin user with notif 'xxx deleted this message'

The message is not deleted for everyone

## ISSUE 7. The error 'null is not an object' is shown and the message is not removed for both users if admin removes the message sent by non-admin

**Preconditions:**
1. User_A is the admin of the created group chat
2. User_B is invited into the group chat

**Steps to reproduce:**
1. Send the message as User_B
2. Remove the message as User_A using 'delete for all' option
3. Check the message as User_B
4. Reopen the chat as User_A

**Actual result:**
The error 'null is not an object'  is shown and the message is not removed for users in the group chat

https://user-images.githubusercontent.com/52490791/219428880-0d578c93-9268-48c4-95a4-2fa3fd18507e.mp4

**Expected result:**
The message is removed for both users if admin removes it the group chat

**Logs:**
[Status.log](https://github.com/status-im/status-mobile/files/10758273/Status.log)
https://www.dropbox.com/s/k91zpla3e9s7flh/geth.log?dl=0

**Comment by VolodLytvynenko** [comment]
@yqrashawn Could you help me with one question, please? 
As I noticed in desktop Status there is an option to set permissions to pin messages for all users in the community, but when this permission is enabled then non-admin (MOBILE) users still cannot pin/unpin messages. Should this work for now or this is not in the scope of PR?


image

**Comment by yqrashawn** [comment]
@VladimrLitvinenko not in the scope of this PR, created an issue https://github.com/status-im/status-mobile/issues/15112 
The key is already existing in mobile `pinMessageAllMembersEnabled`, will implement this next

all above issues fixed, ready for review again

**Comment by VolodLytvynenko** [comment]
Hi @yqrashawn Issues 4,5,6,7 are fixed. PR looks good. Thank you for your great work. Could you merge it, please?

**Pull request merged**